### ISSUE TITLE
string builtin handling of empty regex matches is unexpected

### ISSUE THREAD

**Issue by kopischke** [fish-shell/fish-shell#3116]
Using the `string` builtin with a Perl compatible regex using a `<char>*` pattern will return a zero length match first even if there is a match for `<char>`.
### Reproduction Steps:
1. `echo foo | string -r 'o*'`
### Expected behavior:

Prints “oo”.
### Observed behavior:

Prints a zero width match.
### Additional information:

~~`echo foo | string -ra 'o*'` prints two matches, one zero width, one “oo”. The fish documentation clearly says regexes are Perl compatible, and that means [the `*` quantifier should be greedy, i.e. match the longest possible match](http://perldoc.perl.org/perlre.html#Regular-Expressions).~~

[First reported on gitter by Aaron Gyes](https://gitter.im/fish-shell/fish-shell?at=5752eba572e5509b40787d79).

EDIT: as it turns out, this behaviour is actually both [technically correct](https://github.com/fish-shell/fish-shell/issues/3116#issuecomment-224425924) and [in line with regex implementations in major programming languages](https://github.com/fish-shell/fish-shell/issues/3116#issuecomment-224418023), including Perl. However, [`grep` does not report zero width matches](https://github.com/fish-shell/fish-shell/issues/3116#issuecomment-224448381) which leads to the expectation that interactive shell usage would differ from the above.

---
- Fish version: 2.3
- Operating system: OS X 10.11.5, Homebrew install
- Terminal or terminal emulator: Terminal.app

**Comment by floam** [comment]
>echo foo | string -ra 'o*' prints two matches, one zero width, one “oo”.

It prints, 3, a zero width, oo, another zero width:

```
➤ echo foo | string match -r -a 'o*'

oo

floam@AaronPad:~
➤ echo foo | string match -r -a -n 'o*'
1 0
2 2
4 0
```

Playing with pcre2test and different options:
```
pcre2test
PCRE2 version 10.21 2016-01-12
  re> /o*/
data> foo
 0: 
data> foo\=notempty_atstart
 0: oo
data> 
  re> /o*/g
data> foo
 0: 
 0: oo
 0: 
data> foo\=ph
 0: 
Partial match: oo
```

**Comment by kopischke** [comment]
Well, whatever PCRE does, it’s not in line with the actual specification of the `*` quantifier, so it should be fixed in my book.

**Comment by floam** [comment]
While looking into `PCRE2_NOTEMPTY_ATSTART` I found out we already are using that. Our code roughly is mimicking [pcre2demo.c](http://www.pcre.org/current/doc/html/pcre2demo.html). I think we're failing to properly dig deeper after finding the empty match like they describe here (maybe a regression?)

```
* If the previous match WAS for an empty string, we can't do that, as it *
* would lead to an infinite loop. Instead, a call of pcre2_match() is    *
* made with the PCRE2_NOTEMPTY_ATSTART and PCRE2_ANCHORED flags set. The *
* first of these tells PCRE2 that an empty string at the start of the    *
* subject is not a valid match; other possibilities must be tried. The   *
* second flag restricts PCRE2 to one match attempt at the initial string *
* position. If this match succeeds, an alternative to the empty string   *
* match has been found, and we can print it and proceed round the loop,  *
* advancing by the length of whatever was found. If this match does not  *
* succeed, we still stay in the loop, advancing by just one character.   *
* In UTF-8 mode, which can be set by (*UTF) in the pattern, this may be  *
* more than one byte.  
```

**Comment by floam** [comment]
(I don't see how matched=0 is going to happen here to force -a: https://github.com/fish-shell/fish-shell/blob/8d6735cb4194094b6f3e41573bbfffeb3d49d6ea/src/builtin_string.cpp#L459. I see we are printing newlines when they are not necessary as well.)

**Comment by floam** [comment]
Actually, looks like this is handled the same when originally introduced. Any thoughts, @msteed?

**Comment by msteed** [comment]
My intent was to make use of `pcre2_match()` the same way that `pcre2demo.c` does. It looks like fish is doing the right thing by that measure.

Testing on http://regex101.com/ with pcre, javascript, and python regular expressions reports the same 3 matches: empty-at-start, `oo`, and empty-at-end. The empty matches are reported even though the explanation indicates that the `*` quantifier is greedy. The fact that all three engines agree tells me that fish's behavior is correct.

**Comment by kopischke** [comment]
I just tested against actual regex engines in JavaScript (WebKit ES6 subset), Ruby (2.3), Python (2.7) and Perl (5.18) and they behave as described by @msteed. What makes me go up the walls is that, conceptually, this implementation goes against the definition of a greedy quantifier. Quotes (emphasis mine):

>  By default, a quantified subpattern is "greedy", that is, it will match **as many times as possible** (given a particular starting location) while still allowing the rest of the pattern to match.
– [Perl documentation](http://perldoc.perl.org/perlre.html#Regular-Expressions)

> Causes the resulting RE to match 0 or more repetitions of the preceding RE, **as many repetitions as are possible**.
– [Python documentation](https://docs.python.org/2.7/library/re.html?highlight=re#module-re)

Where the ~~expletive deleted~~ does it say “prefer the leftmost match, whatever its length” for an unanchored pattern? This kind of stuff makes me want to go back to vellum and ink…

Anyway, it seems `fish`’s behaviour is, if not correct, at least in line with all other PCRE implementations. Closing, albeit reluctantly.

**Comment by msteed** [comment]
@kopischke: I agree that there is a disconnect between what the docs imply and how the code behaves, and I was surprised to find that other implementations behave this way.

@krader1961: care to bring your 30+ years of regex experience to bear and make sense of this? :wink: Also, you suggested that empty matches outside of capturing groups should be ignored; should we pursue this?

**Comment by kopischke** [comment]
@msteed so was I, though I have to admit that thinking this through (now that I have let off my steam) it _is_ correct: after all, one would expect `/o+/` to match the “o” of the first “to” in “to ook or not to ook”, not the “oo” in “ook” despite `+` being greedy, and `o*` _does_ match “” so … Seems it is easy to forget regexes anchor leftmost and proceed from there when zero width matches are in play, 20 years of regex usage be damned :disappointed:

**Comment by krader1961** [comment]
If I were using python or C I wouldn't be surprised that the pattern matches the empty string. But for some inexplicable reason I, and I suspect most people, expect the same pattern in a shell like fish to match the "oo" portion of the string. Unfortunately I don't think we dare change the behavior because a) it would be incompatible with other PCRE2 implementations and b) I'd worry about edge cases we might break by trying to make the behavior more intuitive. So, sadly, relabeling this "will not fix".

**Comment by floam** [comment]
Is it really not practical to do what pcregrep does?

```
echo foo | pcre2grep 'o*'
foo
```

or

```
echo foo | pcre2grep -o 'o*'
oo
```

**Comment by krader1961** [comment]
> Is it really not practical to do what pcre2grep does?

Don't know. It's certainly worth looking at the pcre2grep code to see what it's doing in the case of your second example. It's certainly more in line with what people expect; albeit the behavior is different than what someone coding this in another language would expect.

**Comment by floam** [comment]
Also, GNU grep:
```
➤ echo foo | ggrep 'o*'
foo
➤ echo foo | ggrep -o 'o*'
oo
```
I think we want to be more like grep than like a languages' re module.

**Comment by msteed** [comment]
From the `pcre2grep` man page:

> Patterns that can match an empty string are accepted, but empty string matches are never recognized. An example is the pattern "(super)?(man)?", in which all components are optional. This pattern finds all occurrences of both "super" and "man"; the output differs from matching with "super|man"  when  only the matching  substrings are being shown.

**Comment by msteed** [comment]
@krader1961: I think you are right about users' expectations, and I think it may stem from the behavior of the various flavors of `grep`, all (?) of which seem to share the behavior that empty matches are never reported.

**Comment by floam** [comment]
I think this should, at the least, remain as an enhancement request.

**Comment by kopischke** [comment]
@floam thanks for pointing out the `grep` behaviour, you are right to reopen this. I’ve changed the issue title to better reflect the question at hand.

**Comment by faho** [comment]
Strictly speaking, this would be a breaking change. So I'm retargetting this to next-major.

**Comment by mqudsi** [comment]
I implemented this in 1540ff036215a4ab85f6dfd6f6a1d9cb33c47967 but I'm afraid that it's unlikely that can be merged. The problem is that fundamentally `grep` and `string match -r` are different: fish exposes group capturing via the output of `string match -r` while `grep` does not. In fact, with `grep` it's impossible to handle individual groups, and it'd be required to switch to something like `pcregrep` instead.

`pcregrep` can get away with ignoring zero-width results by default because its api for dealing with captured groups differs from that of fish; `pcregrep` requires that you indicate which captured group you want to print individually via an optional parameter to `-o`.  On the other hand, fish's `string match -r` makes it easier to obtain _all_ captured groups and use the results without rerunning the regex match by emitting a captured group per line (with the entire capture emitted on the first line by default).

If a zero-width result is ignored (or even if only the first zero-width result is ignored), it becomes impossible to index the results by capture group.

As I do not see us dropping that default behavior and breaking compatibility with pretty much every serious use of  `string match -r` (including many in our own codebase), I'm afraid that this is going to have be closed as won't fix.

(The workaround here is very easy, replace `o*` with `o+` because you explicitly _don't_ want the zero-length match.)
